File: src/csharp/Intel/Iced/Intel/DecoderInternal/MvexOpCodeHandlerReader.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if DECODER && MVEX
using System;

namespace Iced.Intel.DecoderInternal {
	sealed class MvexOpCodeHandlerReader : OpCodeHandlerReader {
		public override int ReadHandlers(ref TableDeserializer deserializer, OpCodeHandler?[] result, int resultIndex) {
			ref var elem = ref result[resultIndex];
			switch (deserializer.ReadMvexOpCodeHandlerKind()) {
			case MvexOpCodeHandlerKind.Invalid:
				elem = OpCodeHandler_Invalid.Instance;
				return 1;

			case MvexOpCodeHandlerKind.Invalid2:
				result[resultIndex] = OpCodeHandler_Invalid.Instance;
				result[resultIndex + 1] = OpCodeHandler_Invalid.Instance;
				return 2;

			case MvexOpCodeHandlerKind.Dup:
				int count = deserializer.ReadInt32();
				var handler = deserializer.ReadHandler();
				for (int i = 0; i < count; i++)
					result[resultIndex + i] = handler;
				return count;

			case MvexOpCodeHandlerKind.HandlerReference:
				elem = deserializer.ReadHandlerReference();
				return 1;

			case MvexOpCodeHandlerKind.ArrayReference:
				throw new InvalidOperationException();

			case MvexOpCodeHandlerKind.RM:
				elem = new OpCodeHandler_RM(deserializer.ReadHandler(), deserializer.ReadHandler());
				return 1;

			case MvexOpCodeHandlerKind.Group:
				elem = new OpCodeHandler_Group(deserializer.ReadArrayReference((uint)MvexOpCodeHandlerKind.ArrayReference));
				return 1;

			case MvexOpCodeHandlerKind.W:
				elem = new OpCodeHandler_W(deserializer.ReadHandler(), deserializer.ReadHandler());
				return 1;

			case MvexOpCodeHandlerKind.MandatoryPrefix2:
				elem = new OpCodeHandler_MandatoryPrefix2(deserializer.ReadHandler(), deserializer.ReadHandler(), deserializer.ReadHandler(), deserializer.ReadHandler());
				return 1;

			case MvexOpCodeHandlerKind.EH:
				elem = new OpCodeHandler_EH(deserializer.ReadHandler(), deserializer.ReadHandler());
				return 1;

			case MvexOpCodeHandlerKind.M:
				elem = new OpCodeHandler_MVEX_M(deserializer.ReadCode());
				return 1;

			case MvexOpCodeHandlerKind.MV:
				elem = new OpCodeHandler_MVEX_MV(deserializer.ReadCode());
				return 1;

			case MvexOpCodeHandlerKind.VW:
				elem = new OpCodeHandler_MVEX_VW(deserializer.ReadCode());
				return 1;

			case MvexOpCodeHandlerKind.HWIb:
				elem = new OpCodeHandler_MVEX_HWIb(deserializer.ReadCode());
				return 1;

			case MvexOpCodeHandlerKind.VWIb:
				elem = new OpCodeHandler_MVEX_VWIb(deserializer.ReadCode());
				return 1;

			case MvexOpCodeHandlerKind.VHW:
				elem = new OpCodeHandler_MVEX_VHW(deserializer.ReadCode());
				return 1;

			case MvexOpCodeHandlerKind.VHWIb:
				elem = new OpCodeHandler_MVEX_VHWIb(deserializer.ReadCode());
				return 1;

			case MvexOpCodeHandlerKind.VKW:
				elem = new OpCodeHandler_MVEX_VKW(deserializer.ReadCode());
				return 1;

			case MvexOpCodeHandlerKind.KHW:
				elem = new OpCodeHandler_MVEX_KHW(deserializer.ReadCode());
				return 1;

			case MvexOpCodeHandlerKind.KHWIb:
				elem = new OpCodeHandler_MVEX_KHWIb(deserializer.ReadCode());
				return 1;

			case MvexOpCodeHandlerKind.VSIB:
				elem = new OpCodeHandler_MVEX_VSIB(deserializer.ReadCode());
				return 1;

			case MvexOpCodeHandlerKind.VSIB_V:
				elem = new OpCodeHandler_MVEX_VSIB_V(deserializer.ReadCode());
				return 1;

			case MvexOpCodeHandlerKind.V_VSIB:
				elem = new OpCodeHandler_MVEX_V_VSIB(deserializer.ReadCode());
				return 1;

			default:
				throw new InvalidOperationException();
			}
		}
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/dec/MvexOpCodeHandlerReader.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.dec;

import com.github.icedland.iced.x86.internal.dec.MvexOpCodeHandlerKind;

final class MvexOpCodeHandlerReader extends OpCodeHandlerReader {
	@Override
	int readHandlers(TableDeserializer deserializer, OpCodeHandler[] result, int resultIndex) {
		OpCodeHandler elem;
		switch (deserializer.readMvexOpCodeHandlerKind()) {
		case MvexOpCodeHandlerKind.INVALID:
			elem = OpCodeHandler_Invalid.Instance;
			break;

		case MvexOpCodeHandlerKind.INVALID2:
			result[resultIndex] = OpCodeHandler_Invalid.Instance;
			result[resultIndex + 1] = OpCodeHandler_Invalid.Instance;
			return 2;

		case MvexOpCodeHandlerKind.DUP:
			int count = deserializer.readInt32();
			OpCodeHandler handler = deserializer.readHandler();
			for (int i = 0; i < count; i++)
				result[resultIndex + i] = handler;
			return count;

		case MvexOpCodeHandlerKind.HANDLER_REFERENCE:
			elem = deserializer.readHandlerReference();
			break;

		case MvexOpCodeHandlerKind.ARRAY_REFERENCE:
			throw new UnsupportedOperationException();

		case MvexOpCodeHandlerKind.RM:
			elem = new OpCodeHandler_RM(deserializer.readHandler(), deserializer.readHandler());
			break;

		case MvexOpCodeHandlerKind.GROUP:
			elem = new OpCodeHandler_Group(deserializer.readArrayReference(MvexOpCodeHandlerKind.ARRAY_REFERENCE));
			break;

		case MvexOpCodeHandlerKind.W:
			elem = new OpCodeHandler_W(deserializer.readHandler(), deserializer.readHandler());
			break;

		case MvexOpCodeHandlerKind.MANDATORY_PREFIX2:
			elem = new OpCodeHandler_MandatoryPrefix2(deserializer.readHandler(), deserializer.readHandler(), deserializer.readHandler(), deserializer.readHandler());
			break;

		case MvexOpCodeHandlerKind.EH:
			elem = new OpCodeHandler_EH(deserializer.readHandler(), deserializer.readHandler());
			break;

		case MvexOpCodeHandlerKind.M:
			elem = new OpCodeHandler_MVEX_M(deserializer.readCode());
			break;

		case MvexOpCodeHandlerKind.MV:
			elem = new OpCodeHandler_MVEX_MV(deserializer.readCode());
			break;

		case MvexOpCodeHandlerKind.VW:
			elem = new OpCodeHandler_MVEX_VW(deserializer.readCode());
			break;

		case MvexOpCodeHandlerKind.HWIB:
			elem = new OpCodeHandler_MVEX_HWIb(deserializer.readCode());
			break;

		case MvexOpCodeHandlerKind.VWIB:
			elem = new OpCodeHandler_MVEX_VWIb(deserializer.readCode());
			break;

		case MvexOpCodeHandlerKind.VHW:
			elem = new OpCodeHandler_MVEX_VHW(deserializer.readCode());
			break;

		case MvexOpCodeHandlerKind.VHWIB:
			elem = new OpCodeHandler_MVEX_VHWIb(deserializer.readCode());
			break;

		case MvexOpCodeHandlerKind.VKW:
			elem = new OpCodeHandler_MVEX_VKW(deserializer.readCode());
			break;

		case MvexOpCodeHandlerKind.KHW:
			elem = new OpCodeHandler_MVEX_KHW(deserializer.readCode());
			break;

		case MvexOpCodeHandlerKind.KHWIB:
			elem = new OpCodeHandler_MVEX_KHWIb(deserializer.readCode());
			break;

		case MvexOpCodeHandlerKind.VSIB:
			elem = new OpCodeHandler_MVEX_VSIB(deserializer.readCode());
			break;

		case MvexOpCodeHandlerKind.VSIB_V:
			elem = new OpCodeHandler_MVEX_VSIB_V(deserializer.readCode());
			break;

		case MvexOpCodeHandlerKind.V_VSIB:
			elem = new OpCodeHandler_MVEX_V_VSIB(deserializer.readCode());
			break;

		default:
			throw new UnsupportedOperationException();
		}
		result[resultIndex] = elem;
		return 1;
	}
}
